Fields difference

manycoding opened the pull request:
Implements #158 

I wrote a feature which we already had with price rules, which mimics set intersection and difference. It still reads `product_url_field` & `name_field` tags, but the output is different:

How it looks now:
![Screenshot 2019-09-13 at 12 56 41](https://user-images.githubusercontent.com/10396557/64876658-00152d80-d626-11e9-9620-65e429624370.png)

To compare the difference (notice the change in the output format) and timing - https://jupyter.scrapinghub.com/user/u/lab/tree/shared/Experiments/Arche/PRs/difference.ipynb

1. Should we `lower().strip()` strings before comparing? Should we normalize numbers (e.g. 5 = 5.0). In this rule, should we care about particular format (case matters) or more about the meaning?

ejulio commented on src/arche/rules/compare.py:
```diff
@@ -0,0 +1,74 @@
+            new = source[~(source.isin(target))]
+            missing = target[~(target.isin(source))]
+        except SystemError:
+            source = source.apply(str)
```
why not `astype`?

manycoding replied:
Right, astype is a bit faster

ejulio commented on src/arche/rules/compare.py:
```diff
@@ -0,0 +1,74 @@
+        if len(missing) == 0:
+            continue
+
+        if len(missing) < 6:
```
This `6` here is quite magical.
maybe a constant to make its intent clear

simoess commented:
> 1. Should we lower().strip() strings before comparing? Should we normalize numbers (e.g. 5 = 5.0). In this rule, should we care about particular format (case matters) or more about the meaning?

I think that would be nice to be add an option to be able to apply a transformation to some column values. But it shouldn't be the default rule behavior to normalize strings/numbers.

Also, probably it would be useful to be able to access the values that are in the intersection/difference.

manycoding commented:
I'll leave normalization for the other pr.

>Also, probably it would be useful to be able to access the values that are in the intersection/difference.

One can extract keys and then values:
`a.target_items.df.loc[list(res.messages[Level.ERROR][0].errors.values())[0]]`

But that's not very conveniet. In what form would you like to see values? There's an option to shorten it, I think it could be:
* missing values - [value1, ...., value20]
* missing value1 - [key1, ...., keyx]
... missing valuex - [key1, ...., keyj]
* or like this with pandas loc

simoess commented:
> I'll leave normalization for the other pr.
> 
> > Also, probably it would be useful to be able to access the values that are in the intersection/difference.
> 
> One can extract keys and then values:
> `a.target_items.df.loc[list(res.messages[Level.ERROR][0].errors.values())[0]]`
> 
> But that's not very conveniet. In what form would you like to see values? There's an option to shorten it, I think it could be:
> 
> * missing values - [value1, ...., value20]
> * missing value1 - [key1, ...., keyx]
>   ... missing valuex - [key1, ...., keyj]
> * or like this with pandas loc

The first option. It's easy to manipulate the results after. 
.missing_values -> [value1, ..., value20]

Also I think that having all missing items on a df would be useful(e.g checking urls) 
.missing_values_items -> pandas df with the complete item information

manycoding commented:
>Also I think that having all missing items on a df would be useful(e.g checking urls)
.missing_values_items -> pandas df with the complete it

Can you give an example? Do you mean we should find added\dropped values between job by default or that we should have a feature which does that?

simoess commented:
> > Also I think that having all missing items on a df would be useful(e.g checking urls)
> > .missing_values_items -> pandas df with the complete it
> 
> Can you give an example? Do you mean we should find added\dropped values between job by default or that we should have a feature which does that?

The rule could return the items that are in the set difference/intersection between two jobs as a pandas df, or an easy way to access these values. Sometimes missing values needs to be further investigated, to check if they really should be missing.

manycoding commented:
> The rule could return the items that are in the set difference/intersection between two jobs as a pandas df, or an easy way to access these values. Sometimes missing values needs to be further investigated, to check if they really should be missing.

Got it, I created #169. It differs from this feature as here we care about values only, and there we care if any items changed.

manycoding commented:
I added errors by introducing `more_stats` field, also normalization flag.
```
res = arche.rules.compare.fields(
    a.source_items.df, a.target_items.df, ["product_name", "url"]
)
res.more_stats["product_name"]["new"]
```
So for this rule `more_stats` is `{field: {new\same\missing: pd.Series}}`

I also simplified tests by writing an assert and removing unneccesary logic from the code, there're lots of changes.

Please review @simoess @ejulio

manycoding commented on tests/conftest.py:
```diff
@@ -217,30 +218,50 @@ def create_result(
-            elif left_v != right_v:
-                assert_msgs.extend([f"{left_v}", "!=", f"{right_v}"])
-        return assert_msgs
+def assert_results_equal(left: Result, right: Result, **kwargs):
```
@victor-torres This is the assert I told about. Much less code now.

simoess commented on src/arche/rules/compare.py:
```diff
@@ -0,0 +1,94 @@
+        else:
+            msg = f"{', '.join(missing.unique()[:5].astype(str))}..."
+        msg = f"{msg} `{field}s` are missing"
+        if len(missing) / len(target_df) >= err_thr:
```
It's unclear what err_thr does. I think it would be a good idea to document what the variable does on the doctring of the function;
